File: backend/src/main/kotlin/io/shelters/ShelterCountryIso.kt
package io.shelters

import kotlin.math.abs

/**
 * Maps coordinates to ISO 3166-1 alpha-2 using rough bounding boxes — **no Mapbox Tilequery**
 * (zero billable geocoding HTTP calls). Borders may classify ambiguous points incorrectly.
 * Regions are checked in order (specific overlaps handled first).
 */
internal fun countryIsoFromLatLon(
    latitude: Double,
    longitude: Double,
): String? {
    if (!latitude.isFinite() || !longitude.isFinite()) return null
    if (abs(latitude) > 90.0 || abs(longitude) > 180.0) return null
    for ((minLat, maxLat, minLon, maxLon, iso) in COUNTRY_REGION_BOXES) {
        if (latitude in minLat..maxLat && longitude in minLon..maxLon) return iso
    }
    return null
}

private val COUNTRY_REGION_BOXES: List<RegionBox> =
    listOf(
        /** Croatia (pilot shelter outside NL). */
        RegionBox(42.2, 46.95, 13.0, 19.55, "HR"),
        /** Netherlands (directory pilot). */
        RegionBox(50.75, 53.55, 3.3, 7.23, "NL"),
    )

private data class RegionBox(
    val minLat: Double,
    val maxLat: Double,
    val minLon: Double,
    val maxLon: Double,
    val iso: String,
)

File: frontend/src/map/shelterCountryIso.ts
/**
 * Same bounding regions as backend `ShelterCountryIso.kt` — keeps map tint when `countryCode`
 * is absent from the API (older deploy) and avoids duplicate Tilequery calls.
 */
const REGION_BOXES: readonly [
  minLat: number,
  maxLat: number,
  minLon: number,
  maxLon: number,
  iso: string,
][] = [
  [42.2, 46.95, 13.0, 19.55, 'HR'],
  [50.75, 53.55, 3.3, 7.23, 'NL'],
]

export function countryIsoFromLatLon(latitude: number, longitude: number): string | null {
  if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) return null
  if (Math.abs(latitude) > 90 || Math.abs(longitude) > 180) return null
  for (const [minLat, maxLat, minLon, maxLon, iso] of REGION_BOXES) {
    if (latitude >= minLat && latitude <= maxLat && longitude >= minLon && longitude <= maxLon) {
      return iso
    }
  }
  return null
}
